Fix list_nodes in the Azure arm driver

jan-mue opened the pull request:
## Fix the list_nodes function in the Azure ARM driver

### Description

The Azure API returns an attribute `nextLink` for pagination that is currently not used in the Azure ARM driver.
Documentation: https://learn.microsoft.com/en-us/rest/api/compute/virtual-machines/list-all?tabs=HTTP

This PR fixes the issue where not all VMs are returned for subscriptions with a lot of VMs (#1824)

### Status

ready for review

### Checklist (tick everything that applies)

- [ ] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [ ] Documentation
- [ ] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)
- [ ] [ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes) (required for bigger changes)

jan-mue committed:
fix list_nodes for the azure arm driver

Kami commented:
Thanks for the contribution.

It would be great if you could also add a unit test for this change.

mohammad-aburadeh committed:
Support all S3 storage classes

mohammad-aburadeh committed:
Add test case in test_s3.py to test upload object with storage class

cosmo-ray committed:
outscale: fix list_volumes

Signed-off-by: Matthias Gatto <matthias.gatto@outscale.com>

cosmo-ray committed:
outscale: fix private_ips and public_ips from nodes

Signed-off-by: Matthias Gatto <matthias.gatto@outscale.com>

llamasoft committed:
Fix exception when using internal URL

Cloud Files has no concept of an "internal" CDN so the endpoint type must be hard-coded to "external".
Sample Cloud Files service catalog listing: https://docs.rackspace.com/docs/cloud-files/v1/getting-started/authenticate#cloud-files-auth-response

jan-mue committed:
add unit tests

jan-mue commented:
@Kami sorry for the long delay, but I've added some unit tests now. Could you please have another look at the PR?

jan-mue committed:
Merge branch 'trunk' into fix-azure-list-nodes

cosmo-ray committed:
fix crash if no PublicIp or no PrivateIp

Kami committed:
Fix failing test due to extra , which made argument a tuple, fix lint
and reformat the code.

Kami committed:
Add changelog entry for #1875.

Kami committed:
Upgrade rstcheck.

Kami committed:
Re-generate supported method tables.

Kami committed:
Upgrade codecov dependency.

Kami committed:
Switch to non-deprecated GHA action based codecov coverage uploader.

Kami committed:
Update dependabot config.

Kami committed:
Upgrade pytest dependencies.

Kami committed:
Upgrade requests and requests_mock library.

Kami committed:
Upgrade various dev and tests dependencies.

Kami committed:
Squashed '.github/actions/gh-action-pip-audit/' changes from 9075e938d..d499194be

d499194be README: prep 1.0.8
cf52d21d8 Remove pin on requests (fixes #41) (#42)
3ac8fed01 README: prep 1.0.7 (#40)
f7e969538 requirements: constrain requests below 2.30 (#39)
28aa5e1be New issue templates (#37)
1abec09c8 action: replace `internal-be-careful-debug` (#36)
75edeacda README: prep 1.0.6 (#34)
666b1b883 Check that output file exists before opening (#33)

git-subtree-dir: .github/actions/gh-action-pip-audit
git-subtree-split: d499194be74aeb3bc7dbed3a224a87e1831132c7

Kami committed:
Merge commit '21f6883472df778d022fd9ffe34c442d18fa52be' into trunk

Kami committed:
Update GHA actions.

Kami committed:
Upgrade black.

Kami committed:
Reformat code with the latest version of black library.

Kami committed:
Ignore setuptools vulnerability over we don't have a direct control (we
just use version which is installed in the end user environment).

Kami committed:
Fix codecov coverage upload CI step.

Kami committed:
Fix conflicing dependency version.

Kami committed:
Temporary disable PyPy 3.7 tests which are failing on CI.

Kami committed:
Merge branch 'trunk' of https://github.com/outscale-mgo/libcloud into outscale-mgo-trunk

Kami committed:
Add changelog entry for #1877.

Kami committed:
Merge branch 'patch-1' of https://github.com/llamasoft/libcloud into llamasoft-patch-1

Kami committed:
Add test case for get_endpoint() and use_internal_url=True.

Kami committed:
Add changelog entry for #1883 and #1884.

Kami commented:
@jan-mue Thanks for adding a test case. I will review it shortly and if everything looks fine, merge changes into trunk.

Kami commented on libcloud/compute/drivers/azure_arm.py:
```diff
@@ -468,11 +468,17 @@ def list_nodes(self, ex_resource_group=None, ex_fetch_nic=True, ex_fetch_power_s
-        ]
+        params = {"api-version": VM_API_VERSION}
+        nodes = []
+        while True:
```
I think this code is far for now, but to be on the safe side and avoid possible infinite loop, I think it would be good to add additional guard here - e.g. bail out if we are still stuck in this loop for more than X seconds.

Kami committed:
Merge branch 'fix-azure-list-nodes' of https://github.com/jan-mue/libcloud into jan-mue-fix-azure-list-nodes

Kami commented:
@jan-mue Can you please also sign an ICLA when you get a chance (https://libcloud.readthedocs.io/en/latest/development.html#contributing-bigger-changes)? Thanks.

Kami committed:
Update the code so we don't end up in an infinite loop in case API
returns invalid response or similar.

Also add a test case for it.

Kami commented:
I pushed a change so we don't end up in an infinite loop in case server returns a bad response, there is a bug in the code or similar - c49ecd0a4b8fe1e3820d8764f9aded21e5dee11c.

I picked that 60 seconds limit arbitrarily, but if needed or wanted, it can be changed (e.g. increased) - @jan-mue Can you please advise if 60 seconds is enough or the limit should be higher?

I imagine we may want a higher limit since depending on the method arguments, each ``_to_node()`` call will result in additional HTTP request(s).

For now, in case deadline is reached we simply return, but we may also want to throw or at least log a warning in such scenario.

Kami committed:
Bump the pagination limit to 300s.

Kami commented:
EDIT: I went ahead and bumped that limit to 300 seconds for now. This way I can wrap up and merge this PR.

And then based on your future feedback @jan-mue, we can adjust this value.

Kami committed:
Add changelog entry for #1824 #1850.

Kami committed:
Update the comment.

Kami committed:
Fix codecov config.

Kami committed:
Merge branch 'trunk' into fix-azure-list-nodes
